### ISSUE TITLE
Heavy load with large GIFs eventually make the app unresponsive without an error

### ISSUE THREAD

**Issue by Robert-Wett** [lovell/sharp#225]
I've been using your library to power all of our sites thumbnailing/image resizing operations, and have been running into a lot of issues lately under heavy load, specifically with GIFs. We are running `^0.9.1` sharp, and using `vips-8.0.1` for the GIF functionality.

The issue is that when a server using sharp has 5+ or so large (9+ MB, sharp.cache set to 0, but it exhibited the same behavior with defaults ) GIF's it needs to process, it will at some point (sometimes, not always) just become unresponsive and no longer respond to any requests without throwing any kind of error. I am having to `.toBuffer` these jobs before I call `.toFile`, but not sure if that matters if I'm not using streams. The worst part is that it cripples all functionality, but the box is still somehow able to maintain it's consumer connection to rabbitmq and will still pull messages and hoard them until manually restarted (that part just sucks, not expecting you to know why that happens :smile: ). We are in the process of setting up load testing environments so that we can play around with `sharp.concurrency` and `UV_THREADPOOL_SIZE` to see if we can eliminate this behavior as a possibility, but man would it be nice if there were a way to throw an error when it won't be able to read in a large GIF/file to memory (or w/e is crashing it) so I could reject the job and send it off to another box that has more memory available.  

Thanks for your work!

**Comment by lovell** [comment]
Hi Rob, libvips relies on either ImageMagick or GraphicsMagick for processing GIF files, both of which end up decompressing the entire image into memory.

I've found GraphicsMagick to be a bit more reliable in terms of memory consumption - you can force libvips to use it at compile time via the `./configure --with-magickpackage=GraphicsMagick ...` flag.

Using sharp's `toFile` instead of `toBuffer` should also reduce memory consumption (Streams use Buffers under the hood).

If you're using Linux, is the kernel's out-of-memory killer kicking in (`dmesg` can confirm this)? If so, perhaps limit the memory for your process so it will fail to allocate memory and generate an error rather than be suspended.

I don't know how much control you have over the source images, but if you can influence these then for line art diagrams I recommend (non interlace) PNG for the lowest CPU and memory requirements.

**Comment by lovell** [comment]
@Robert-Wett Have you been able to make any progress with this problem?

**Comment by Robert-Wett** [comment]
Hey @lovell! Thanks for being so responsive.
Unfortunately, cursory experiments with the usual settings didn't seem to yield any better results - it's really easy to run into an OOM error with GIFs. I haven't yet rebuilt the libvips package with that flag, so if you think there is a significant gain to be made from that vs. using an Image Magick library that pulls the first frame of the gif and processes that then let me know ( I've since written a GM/IM wrapper ( node gm's resizing method when it's just given a width just doesn't work, so I've had to `subClass` IM) to handle all GIF operations for now). The thing is, sharp just doesn't seem to be able to let go of memory for me unless it absolutely has to - it's super pronounced with GIF's because it will actually clear out memory to read it in, and you can see a drastic drop in used memory - but for a constant stream of smaller images, it just hovers at the danger zone and eventually runs OOM

**Comment by lovell** [comment]
If you're able to control the version of ImageMagick in use then check it against [the changelog](http://www.imagemagick.org/script/changelog.php) as I see a few recent memory leak fixes.

Any leaks are still present when spawning a per-image `convert` process but are masked as the OS can reclaim the heap as each process ends.

If you've not already done so, you might want to experiment with the V8 garbage collector options (e.g. max old space) if you're seeing the `node` process cling onto what could be fragmented memory. Buffer objects in Node v0.10 are especially prone to this.

**Comment by Robert-Wett** [comment]
Hey @lovell, just wanted to follow up and let you know that this was *mostly* due to just over saturating the servers with jobs in parallel, and I had to set up a `prefetch` value for my consumers connecting to the rabbit queue to help throttle it a bit. With this setting I'm seeing way more expected behavior.

**Comment by lovell** [comment]
Thanks Rob, I'm glad to hear you've been able to solve (some of) the problem.

If you're able to share details publicly, it's always interesting to learn more about what people are using this module for over at the survey - see #35

**Comment by lovell** [comment]
@Robert-Wett You may be interested in the fix for a libvips memory leak relating to animated GIFs - see https://github.com/jcupitt/libvips/pull/313

**Comment by Robert-Wett** [comment]
Thanks so much for the update!

**Comment by automata** [comment]
@lovell I'm reading some closed issues regarding memory leaks and I wonder: shouldn't sharp just read the first frame of GIFs using ImageMagick or GM?

**Comment by lovell** [comment]
@automata Yes, see #369
